apollo-federation/src/sources/connect/validation/entity/keys.rs

```diff
@@ -0,0 +1,215 @@
+                        directive.argument_by_name(&FEDERATION_FIELDS_ARGUMENT_NAME, schema).ok().and_then(|arg| arg.as_str()).unwrap_or_default(),
+                        type_name,
+                    ),
+                    locations: directive
+                        .line_column_range(&schema.sources)
+                        .into_iter()
+                        .collect(),
+                });
+            }
+        }
+
+        messages
+    }
+}
+
+impl std::fmt::Debug for EntityKeyChecker<'_> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.debug_struct("EntityKeyChecker")
+            .field(
+                "resolvable_keys",
+                &self
+                    .resolvable_keys
+                    .iter()
+                    .map(|(fs, _, _)| {
+                        format!(
+                            "... on {} {}",
+                            fs.selection_set.ty,
+                            fs.selection_set.serialize().no_indent()
+                        )
+                    })
+                    .collect_vec(),
+            )
+            .field(
+                "entity_connectors",
+                &self
+                    .entity_connectors
+                    .values()
+                    .flatten()
+                    .map(|fs| {
+                        format!(
+                            "... on {} {}",
+                            fs.selection_set.ty,
+                            fs.selection_set.serialize().no_indent()
+                        )
+                    })
+                    .collect_vec(),
+            )
+            .finish()
+    }
+}
+
+/// Given the parts of the connector that contain variables, synthesize a field
+/// set appropriate for use in a @key directive.
+///
+/// This only looks at `$args` variables because those are the ones used
+/// for `entity: true` connectors.
+///
+/// TODO: this is simpler to code in expand/mod.rs and might be something we
+/// could combine. The expand code not only generates a key, but also copies
+/// the types related to the key into the expanded subgraph, so it does more work
+/// than this function.
+pub(crate) fn make_key_field_set_from_variables(
+    schema: &Schema,
+    object_type_name: &Name,
+    variables: &[VariableReference<Namespace>],
+    resolver: EntityResolver,
+) -> Result<Option<Valid<FieldSet>>, Message> {
+    let params = variables
+        .iter()
+        .filter(|var| match resolver {
+            EntityResolver::Explicit => var.namespace.namespace == Namespace::Args,
+            EntityResolver::Implicit => var.namespace.namespace == Namespace::This,
+        })
+        .unique()
+        .collect_vec();
+
+    if params.is_empty() {
+        return Ok(None);
+    }
+
+    let mut keys = Vec::with_capacity(params.len());
+    for VariableReference { path, .. } in params {
```

Comment:
feedback from @nicholascioli — let's merge the variables here (`$args.a.b.c` + `$args.a.b.c` = `a { b { c d } }`) and then we can get rid of the optimize_field_sets fn entirely!